code/globalincs/undosys.h

```diff
@@ -0,0 +1,264 @@
+	T* container;   //!< Optional reference to this item's container
+};
+
+
+/*!
+ * @brief Class which handles multiple undo operations as a single op within the Undo_system
+ */
+class Undo_stack : public Undo_item_base
+{
+public:
+	Undo_stack();
+
+	Undo_stack(size_t size);
+
+	~Undo_stack();
+
+	/*!
+	 * @brief Restores all items within the undo stack
+	 *
+	 * @details Maintains an internal flag which will reverse direction on the next call, thereby redo-ing
+	 * 
+	 * @returns A pair of references to the last item restored
+	 */
+	std::pair<const void*, const void*> restore();
+
+	/*!
+	 * @brief Saves the item onto the undo stack
+	 *
+	 * @param[in] item      The item to save
+	 * @param[in] container (Optional) The container wherein this item is located
+	 * @note Call this _before_ you do your operation on the item
+	 */
+	template<typename T>
+	size_t save(T& item, T* container = nullptr);
+
+	/*!
+	 * @brief Calls ::reserve() on the internal vector
+	 */
+	void reserve(size_t size);
+private:
+	bool reverse;   // Direction to walk the stack. forward = false, reverse = True
+
+	std::vector<Undo_item_base*> stack;
+};
+
+/*!
+ * @brief Generic Undo/Redo system. Save whatever, restore whatever!
+ *
+ * @details Currently uses a pair of deques. Would ideally use a ring container so that the actual Undo_item instances don't go anywhere.
+ */
+class Undo_system
+{
+public:
+	Undo_system();
+
+	Undo_system(uint _undos);
+
+	/*!
+	 * @brief Deletes all undo and redo data
+	 */
+	void clear();
+	
+	/*!
+	 * @brief Saves the item onto the undo stack
+	 *
+	 * @param[in] item      The item to save
+	 * @param[in] container (Optional) The container wherein this item is located
+	 * @note Call this _before_ you do your operation on the item
+	 */
+	template<typename T>
+	size_t save(T& item, T* container = nullptr);
+
+	/*!
+	 * @brief Saves a stack of undo-items as a single undo-item within the system
+	 */
+	size_t save_stack(Undo_stack& stack);
+
+	/*!
+	 * @brief Undo's the last changed item and save the changes into the Redo stack
+	 *
+	 * @returns A pair of const void* which reference the affected item, and its container (if it was provided)
+	 */
+	std::pair<const void*, const void*> undo();
+
+	/*!
+	 * @brief Redo's the last changed item and save the changes into the undo
+	 *
+	 * @returns A pair of const void* which reference the affected item, and its container (if it was provided)
+	 */
+	std::pair<const void*, const void*> redo();
+
+	/*!
+	 * @brief Returns the size of the undo stack
+	 */
+	size_t size();
+
+	/*!
+	 * @brief Returns the size of the redo stack
+	 */
+	size_t size_redo();
+
+private:
+	uint max_undos;	//!< Max number of Undo's available
+
+	/*! 
+	 * @brief The undo and redo stacks. These are containers of pointers, because we can't simply store the base class.
+	 *   If we did, then the data of the Undo_item instances would be sliced out.
+	 * @{
+	 */
+	SCP_deque<Undo_item_base*> undo_stack;
```

Comment:
You should use a `unique_ptr` if you manage this memory yourself which is, as far as I can tell, the case since you use `delete` on the items in the lists.